[#1405] Modify trait utilities to support prefixed trait keys & add SelectChoices

arbron opened the pull request:
- Modifies a number of utility methods in trait utilities to support prefixed trait keys (`tool:art:potter` instead of just `potter`)
- Introduces the `SelectChoices` type which offers a number of methods for filtering, sorting, cloning, etc. nested sets of options like those represented by tool categories

Fyorl commented on module/documents/actor/trait.mjs:
```diff
@@ -232,24 +369,73 @@ export function keyLabel(trait, key) {
+
+  // Single entry in pool
+  // { count: 3, pool: ["skills:*"] } -> any three skills
+  // { count: 3, pool: ["skills:*"] } (final) -> three other skills
```
I might need a bit of elaboration of how the `final` option is supposed to work. Is it only for the purposes of formatting the list for display?

arbron replied:
Yes, it is to format lists properly when there is a mixture to grants & choices like `"Athletics and three other skills"`, which reads better than `"Athletics and any three skills"`.

Fyorl commented on module/documents/actor/trait.mjs:
```diff
@@ -232,24 +369,73 @@ export function keyLabel(trait, key) {
+ * @param {Set<string>} grants                       Guaranteed trait grants.
+ * @param {TraitChoice[]} [choices=[]]               Trait choices.
+ * @param {object} [options={}]
+ * @param {string} [options.choiceMode="inclusive"]  Choice mode.
```
If there are only a handful of options we can annotate it like this:

```suggestion
 * @param {"inclusive"|"exclusive"} [options.choiceMode="inclusive"]  Choice mode.
```

Fyorl commented on module/documents/actor/trait.mjs:
```diff
@@ -232,24 +369,73 @@ export function keyLabel(trait, key) {
+    sections.push(choiceListFormatter.format(choiceSections));
+  }
+
+  const listFormatter = new Intl.ListFormat(game.i18n.lang, { style: "long", type: "conjunction" });
```
We can use `game.i18n.getListFormatter` which will lazily cache these formatters though it's not really a big deal.

arbron replied:
Hey, new method I didn't know existed, great!

Fyorl reviewed (changes requested):
I can't help but feel like we perhaps wouldn't need to do quite so much work in `Trait` if we had more consistent and predictable structuring in our `CONFIG` as well as for where traits are stored on the actor. I'm not sure there's enough of a tangible benefit to changing it though at this stage given how much churn and breakage it would cause.

It seems that we only really ever have one level of `children` in our current trait configuration. I wonder, then, if having a fully-recursive tree structure and associated operations is necessary, and if there are maybe some simplifications that can be won from recognising this property. It might be that the general tree solution is the simplest though.

In general though I think it's fine that this complexity is tucked away in these utilities. It's an acceptable place to be currently.

If I could ask though for some small examples as to some of the inputs and outputs for the `SelectChoices` class, particularly in regard to how the wildcards work. That would help me a lot with contextualising some of this work. Since it's all sets of strings I'm hoping they won't be too arduous to produce. We could also include them as `@example` annotations.

Some examples for some of the trait methods would also help me, like you have in `choiceLabel`, they were very useful.

arbron commented on module/documents/actor/select-choices.mjs:
```diff
@@ -0,0 +1,191 @@
+   * @param {SelectOptions} other
+   * @returns {SelectOptions}
+   */
+  merge(other) {
```
I can make that change.

This pattern was replicating the syntax of Swift which has a very clear naming to distinguish mutating methods and non-mutating methods unlike the mess of Javascript :).

Fyorl replied:
I guess I can kind of see how the tense of the verb might imply mutation if I think about it for a bit, and if the convention is in place already. That's quite interesting to know about Swift. An `inplace` option is definitely more in-line with the core API conventions though, so thanks for making the change.

arbron commented on module/documents/actor/trait.mjs:
```diff
@@ -17,56 +19,92 @@ const _cachedIndices = {};
- * @param {string} trait  Trait as defined in `CONFIG.DND5E.traits`.
- * @returns {object}      Trait categories defined within `CONFIG.DND5E`.
+ * Get the current trait values for the provided actor.
+ * @param {BlackFlagActor} actor  Actor from which to retrieve the values.
```
Whoops

arbron commented on module/documents/actor/trait.mjs:
```diff
@@ -190,39 +293,73 @@ export function traitIndexFields() {
+ * @returns {string}                Retrieved label.
  */
-export function keyLabel(trait, key) {
+export function keyLabel(key, { count, trait, final }={}) {
```
I think the old way can be handled pretty easily with a deprecation warning in case anyone was using it.

arbron commented:
> I can't help but feel like we perhaps wouldn't need to do quite so much work in `Trait` if we had more consistent and predictable structuring in our `CONFIG` as well as for where traits are stored on the actor. I'm not sure there's enough of a tangible benefit to changing it though at this stage given how much churn and breakage it would cause.

While I think it would be okay to switch things on the actor side, any major change of the config would be a breaking change for any 3rd party content that adds new types.

> It seems that we only really ever have one level of `children` in our current trait configuration. I wonder, then, if having a fully-recursive tree structure and associated operations is necessary, and if there are maybe some simplifications that can be won from recognising this property. It might be that the general tree solution is the simplest though.

We do have the chance of multiple levels when more than one trait type is represented, which mainly comes into play with expertise (e.g. you get a choice between gaining expertise in a any skill or thieves tools).

> If I could ask though for some small examples as to some of the inputs and outputs for the `SelectChoices` class, particularly in regard to how the wildcards work. That would help me a lot with contextualising some of this work. Since it's all sets of strings I'm hoping they won't be too arduous to produce. We could also include them as `@example` annotations.

Added some examples for `filter` and `exclude`.

> Some examples for some of the trait methods would also help me, like you have in `choiceLabel`, they were very useful.

Add a bunch and moved the examples into the JSDocs.

Fyorl commented on module/documents/actor/trait.mjs:
```diff
@@ -84,29 +133,93 @@ export async function choices(trait, chosen=new Set()) {
   }
 
-  return data;
+  const prepareCategory = (key, data, result, prefix) => {
+    let label = _innerLabel(data, traitConfig);
+    if ( !label ) label = key;
+    if ( prefixed ) key = `${prefix}:${key}`;
+    result[key] = {
+      label: game.i18n.localize(label),
+      chosen: chosen.has(key),
+      sorting: traitConfig.sortCategories === true
+    };
+    if ( data.children ) {
+      const children = result[key].children = {};
+      if ( prefixed && any ) {
+        const anyKey = `${key}:*`;
+        children[anyKey] = {
+          label: keyLabel(anyKey).titleCase(),
+          chosen: chosen.has(anyKey), sorting: false, wildcard: true
+        };
+      }
+      Object.entries(data.children).forEach(([k, v]) => prepareCategory(k, v, children, key));
+    }
+  };
```
This is perhaps a silly suggestion, but could this be simplified somewhat by delegating the prefixing to the `SelectChoices` constructor (maybe with an option)? Things started falling into place a lot better once I realised that the children are already prefixed when supplied to the `SelectChoices` constructor. I thought we should maybe document that then, but then I thought we might be able to actually do the prefixing in the constructor, and then it would be obvious.

Maybe that's not feasible, I'm not sure, but maybe you could try it out. Otherwise, ultimately, this is just utility code so it's fine as-is as well with just a small mention in the constructor JSDoc that children are supplied already-prefixed.

arbron replied:
I think the issue is that generating the `chosen` value for each choice requires the fully prefixed key, so if we left prefixing to the `SelectChoices` object we'd probably have to iterate through the entire structure again to set `chosen` or generate the keys twice.

Fyorl replied:
Ah ok, I see. Can we just document the `SelectChoicesEntry` typedef then to mention that the `children` are required to be already-prefixed? I suppose we could add an option to perform the prefixing in `SelectChoices` if we ever needed it. I guess this method is the only place we instantiate a `SelectChoices` object though currently.

arbron replied:
I'll add a note about providing prefixed keys if desired, though we do also use this method with un-prefixed keys in `TraitSelector` so it isn't a requirement.

Fyorl commented on module/documents/actor/trait.mjs:
```diff
@@ -190,39 +303,113 @@ export function traitIndexFields() {
+ *
+ * @example
+ * // Returns "2 other Artisan's Tools"
+ * keyLabel("tool:art:*", { count: 2, final: true });
```
These are super helpful, thank you.

Fyorl commented on module/documents/actor/trait.mjs:
```diff
@@ -232,24 +419,109 @@ export function keyLabel(trait, key) {
+ * choiceLabel({ count: 1, pool: new Set(["tool:thief", "tool:art:*"]) }, { only: true });
+ *
+ * @example
+ * // Returns
```
Don't keep me in suspense like this.

Fyorl commented on module/documents/actor/trait.mjs:
```diff
@@ -232,24 +419,109 @@ export function keyLabel(trait, key) {
+ *   {count: 2, pool: new Set(["languages:*"])}, { count: 1, pool: new Set(["skills:*"])}
+ * ], {choiceMode: "exclusive"});
+ */
+export function localizedList(grants, choices=[], { choiceMode="inclusive" }={}) {
```
Since these seem all kind of optional from the examples, maybe it should just be one big options object:

```suggestion
export function localizedList({ grants=new Set(), choices=[], choiceMode="inclusive" }={}) {
```

Fyorl reviewed (approved):
Looks pretty good to me, thank you. I've left some minor linting comments and some suggestions for consideration, that you can take or leave.

Fyorl reviewed (approved):
Just one comment above about adding documentation then feel free to merge.
